src/ekat/kokkos/ekat_kokkos_utils.hpp

```diff
@@ -53,135 +53,159 @@ namespace impl {
+                           const int& begin,
+                           const int& end,
+                           const Lambda& lambda)
+{
+  ValueType result = Kokkos::reduction_identity<ValueType>::sum();
+  ekat::impl::parallel_reduce<Serialize,ValueType>(team,begin,end,lambda,result);
+  return result;
+}
+
 /*
- * Computes a reduction over a routine described by 'input' for entries 'scalarize(input)([begin,end))'
- * The variable 'input' should be an operator s.t. 'input(k)' returns a Pack. The computed reduction is
- * added to the value 'result'.
+ * Computes a reduction over the range [begin,end) of items provided by the InputProvider
+ * object, which must support calls to operator()(int).
+ * We provide two overloads, one for InputProvider returning a simd-type, and one for
+ * InputProvider returning a non-simd type. In both cases, the result of the reduction
+ * is a non-simd type, and the [begin,end) range refers to the 'scalar' range.
+ * NOTE: for simd-type usage, the simd-type must a) have a specialization of ekat::ScalarTraits
+ *       that sets is_simd=true, b) support operator[](int), and c) have a compile-time size
+ *       equal to the length of the simd array times the size of the individual element type.
+ *       Furtermore, a specialization of Kokkos::reduction_identity<T> must be available for
+ *       the simd type T.
  * NOTE: we do not provide an overload with Serialized defaulted, since this fcn is an impl
  *       detail, and, normally, should not be used by customer apps
  */
-template <bool Serialize, typename TeamMember, typename InputProvider, typename ValueType>
+template<typename InputProvider>
+using ResultType = typename std::remove_cv<
+                      typename std::remove_reference<
+                          decltype( std::declval<InputProvider>()(0) )
+                      >::type
+                   >::type;
+
+template<typename InputProvider>
+using ResultTraits = ekat::ScalarTraits<ekat::impl::ResultType<InputProvider>>;
+
+template <bool Serialize, typename TeamMember, typename InputProvider>
 static KOKKOS_INLINE_FUNCTION
-void view_reduction (const TeamMember& team,
+auto view_reduction (const TeamMember& team,
                      const int begin, // scalar index
                      const int end, // scalar index
-                     const InputProvider& input,
-                     ValueType& result)
+                     const InputProvider& input)
+ -> typename std::enable_if<ekat::impl::ResultTraits<InputProvider>::is_simd,
+                            typename ekat::impl::ResultTraits<InputProvider>::scalar_type>::type
 {
-  using PackType = typename std::remove_reference<decltype(input(0))>::type;
-  constexpr int vector_size = PackType::n;
+  using PackType = ekat::impl::ResultType<InputProvider>;
+  using ValueType = typename ekat::impl::ResultTraits<InputProvider>::scalar_type;
+  constexpr int N = sizeof(PackType) / sizeof(ValueType);
```

Comment:
Is it no longer possible to use PackType::n here?

Reply:
Yes, it is, but I didn't want to unnecessarily rely on the packtype implementation details. This way, the user can potentially swap a different impl (so long as a specialization of `ScalarTraits` exists).